libccp: modify memory limits

deeptir18 opened the pull request:
This commit changes how install expression messages and install update
messages are read in libccp, to remove the restrictions of the Kernel
maximum stack size within a function, and modifies the constants
associated with the maximum fast path program size.

Resolves #17

fcangialosi commented on ccp.c:
```diff
@@ -246,42 +250,48 @@ int ccp_read_msg(
+        state->num_instructions = expr_msg_info.num_instructions;
 
         // parse the expressions 
         for (i=0; i<state->num_expressions; i++) {
```
Could this entire for loop instead just be:

```c
memcpy(state->expressions, msg_ptr, emsg.num_expressions * sizeof(struct ExpressionMsg))
```

?

deeptir18 replied:
Yeah, I think it was this way for debugging purposes but this is more efficient.

deeptir18 replied:
Actually - we can't do this right now. For some reason, portus sends down expression messages in u8s, whereas they are stored as 4 u32s (I don't know if there is a specific reason for this). @akshayknarayan ?

fcangialosi replied:
That’s just how it’s dealt with in the code, it’s the same thing in memory

fcangialosi replied:
Dealing with u8s in rust is equivalent to dealing with chars in C

deeptir18 replied:
I'm not following. If portus sends down 4 u8s, and we say memcpy(state, buffer, sizeof (4 u32s)), it's going to copy stuff past the expression right?

deeptir18 replied:
well, when I tried changing it to  a memcpy, the unit tests didn't pass.

akshayknarayan replied:
@deeptir18 not sure what you mean - https://github.com/ccp-project/portus/blob/master/src/serialize/install.rs#L13 is composed of u32s?

deeptir18 replied:
I'm talking about an Expression Struct itself. If you look in https://github.com/ccp-project/libccp/blob/master/serialize.h#L139, there are 4 u8s in an expression. We can't memcpy this into a struct with 4 u32s directly (as a group) right?

akshayknarayan replied:
I looked more and I see what you are saying, let's discuss offline

fcangialosi commented on ccp.c:
```diff
@@ -246,42 +250,48 @@ int ccp_read_msg(
         }
 
         // parse the instructions
         for (i=0; i<state->num_instructions; i++) {
```
same thing here replacing the loop with a memcpy

deeptir18 replied:
You can't do a memcpy here. If you look at read_instruction(), you actually need to do parsing work (it's not just copying the memory over).

deeptir18 replied:
It works in the expression case because the way they are sent down is exactly how they're laid out in memory in libccp

fcangialosi commented on serialize.c:
```diff
@@ -131,43 +131,43 @@ int write_measure_msg(
     }
 
-    if ((hdr->Len - sizeof(struct CcpMsgHeader)) > sizeof(struct UpdateFieldsMsg)) {
+    memcpy(num_updates, buf, sizeof(u32));
```
this doesn't need to be a memcpy, it's a single value, so you can just set it directly, e.g. `*num_updates = (u32) *buf`

deeptir18 replied:
sure

fcangialosi commented on serialize.h:
```diff
@@ -50,21 +50,22 @@ int serialize_header(char *buf, int bufsize, struct CcpMsgHeader *hdr);
 
 // Some messages contain strings.
-#define  BIGGEST_MSG_SIZE  1024
+#define  BIGGEST_MSG_SIZE  32000
```
32768

fcangialosi commented on serialize.h:
```diff
@@ -50,21 +50,22 @@ int serialize_header(char *buf, int bufsize, struct CcpMsgHeader *hdr);
+#define MAX_INSTRUCTIONS   300 // 50 * 16 = 800 bytes for instructions
+#define MAX_IMPLICIT_REG   6  // fixed number of implicit registers
+#define MAX_REPORT_REG     100 // measure msg 100 * 8 + 4 + 4
+#define MAX_CONTROL_REG    100 // arbitrary
```
can we make this exactly what it supports instead of something arbitrarily lower? no reason to have it 100 if we can support e.g. 110 or something

fcangialosi commented:
TODO: make the expression msg fields u32s (instead of u8s)

fcangialosi commented:
TODO: make the max expressions / instruction msg a kernel module parameter. start with a reasonable default of e.g. 128 or 256 instructions, but allow the user to decide the trade off between per-flow space and program sizes.

fcangialosi commented:
TODO: change everything to match on the portus side, and make sure the portus compiler checks the program size / max instructions / expressions before sending it down as well

akshayknarayan commented on Makefile:
```diff
@@ -1,5 +1,5 @@
 #CC = ${CC} # C compiler
-DEBUG = y
+DEBUG = n
```
separate commit?

akshayknarayan commented on serialize.h:
```diff
@@ -129,60 +130,59 @@ struct __attribute__((packed, aligned(4))) InstructionMsg {
-    u8 num_cond_instrs;
-    u8 event_start_idx;
-    u8 num_event_instrs;
+    u32 cond_start_idx;
```
separate commit

akshayknarayan commented on serialize.h:
```diff
@@ -129,60 +130,59 @@ struct __attribute__((packed, aligned(4))) InstructionMsg {
- * []InstructionMsg: all instructions -> 800 bytes, MAX = 16
- */
-struct __attribute__((packed, aligned(4))) InstallExpressionMsg {
+struct __attribute__((packed, aligned(4))) InstallExpressionMsgHdr {
```
separate commit

akshayknarayan reviewed (approved):
Should probably split into multiple commits.

Commit:
makefile: default is debug off

Commit:
test: remove extra declaration of biggest message

Commit:
libccp: change expression serialization

This commit changes expression structs to be sent down as 4 u32s,
as well as how install and update messages are read to copy directly
from the buffer, to avoid memory limitations with regards to max
stack size in a function.

Commit:
libccp: change limits for per flow state

This commit changes the limits for per flow state to reflect how memory
is only limited by ipc message size. This includes increasing the number
of expressions, instructions and registers available for libccp fast
path programs.
